src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs

```diff
@@ -0,0 +1,153 @@
+                    return value;
+                }
+
+            }
+            else
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache();
+            }
+
+            Session.IncrementRequestCount();
+            value = await Session.DocumentStore.Counters.ForDatabase(Session.DatabaseName).GetAsync(DocId, counter).ConfigureAwait(false);
+            cache.AddOrUptadeCounterValue(counter, value);
+
+            Session.CountersByDocId[DocId] = cache;
+            return value;
+        }
+
+        public async Task<Dictionary<string, long?>> GetAsync(IEnumerable<string> counters)
+        {
+            var countersToGetFromServer = new List<string>();
+            var result = new Dictionary<string, long?>();
+
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                foreach (var counter in counters)
+                {
+                    if (cache.Values.TryGetValue(counter, out var val) || cache.GotAll)
+                    {
+                        result[counter] = val;
+                    }
+                    else
+                    {
+                        countersToGetFromServer.Add(counter);
+                    }
+                }
+
+            }
+            else
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache();
+                countersToGetFromServer = counters.ToList();
+            }
+
+            if (countersToGetFromServer.Count > 0)
+            {
+                Session.IncrementRequestCount();
+
+                var vals = await Session.DocumentStore.Counters.ForDatabase(Session.DatabaseName).GetAsync(DocId, countersToGetFromServer).ConfigureAwait(false);
+                foreach (var kvp in vals)
+                {
+                    cache.AddOrUptadeCounterValue(kvp.Key, kvp.Value);
+                    result[kvp.Key] = kvp.Value;
+                }
+
+                Session.CountersByDocId[DocId] = cache;
+
+            }
+
+            return result;
+        }
+
+        public async Task<Dictionary<string, long>> GetAllAsync()
+        {
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.GotAll)
+                {
+                    var result = new Dictionary<string, long>();
+                    foreach (var kvp in cache.Values)
+                    {
+                        if (kvp.Value.HasValue)
+                        {
+                            result.Add(kvp.Key, kvp.Value.Value);
+                        }
+                    }
+
+                    return result;
+                }
+
+            }
+            else
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache();
+            }
+
+            Session.IncrementRequestCount();
+
+            Dictionary<string, long> values;
+
+            if (cache.MissingCounters != null)
```

Comment:
What is this for?
An in general, if we are already going to the server to get all, we might as well get them all and simplify the code.

Reply:
removed cache.MissingCounters,
if going to server will ask for all